anvil/sample.py

```diff
@@ -25,6 +26,158 @@ def sample(loaded_model, action, target_length: int, n_large=20000) -> torch.Ten
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+
+    Returns
+    -------
+    states[-1]: torch.Tensor
+        the current phi state, for continuity between batches
+    """
+    t_therm = 1000 # ideally come up with a way of working this out on the fly
+
+    states, _ = sample_batch(
+            loaded_model, action, t_therm)
+
+    print(f"Thermalisation: discarded {t_therm} configurations.")
+    return states[-1]
+
+
+def chain_autocorrelation(loaded_model, action, current_state) -> float:
+    r"""
+    Compute an observable-independent measure of the integrated autocorrelation
+    time for the Markov chain.
+
+        \tau_int = 0.5 + sum_{\tau=1}^{\tau_max} \rho(\tau)/\rho(0)
+
+    where \rho(\tau)/\rho(0) is the probability of \tau consecutive rejections,
+    which we estimate by
+
+        \rho(\tau)/\rho(0) = # consecutive runs of \tau rejections / (N - \tau)
+
+    See eqs. (16) and (19) in https://arxiv.org/pdf/1904.12072.pdf
+
+    This measure of autocorrelation is used to provide a first guess for an
+    appropriate subsampling interval,
+
+        sample_interval = ceil(2 * integrated_autocorrelation)
+
+    with the intended effect being that observables on the subsampled chain
+    are entirely decorrelated.
+
+    See http://luscher.web.cern.ch/luscher/lectures/LesHouches09.pdf section 2.2.4
+
+    Parameters
+    ----------
+    loaded_model: Module
+        loaded_model which is going to be used to generate sample states
+    action: Module
+        the action upon which the loaded_model was trained, used to calculate the
+        acceptance condition
+    initial_state:
+        the current state of the Markov chain, after thermalisation
+
+    Returns
+    -------
+    states[-1]: torch.Tensor
+        the current phi state, for continuity between batches
+    sample_interval: float
+        Guess for subsampling interval, based on the integrated autocorrelation time
+
+    """
+    batch_size = 1000 # Hard coded num states for estimating integrated autocorrelation
+
+    # Sample some states
+    states, history = sample_batch(
+            loaded_model, action, batch_size, current_state
+    )
+
+    N = len(history)
+    autocorrelations = torch.zeros(N - 1, dtype=torch.float)
+    consecutive_rejections = 0
+
+    for step in history:
+        if step == True:  # move accepted
+            if consecutive_rejections > 0:  # faster than unnecessarily accessing array
+                autocorrelations[1 : consecutive_rejections + 1] += 1
+            consecutive_rejections = 0
+        else:  # move rejected
+            consecutive_rejections += 1
+    if consecutive_rejections > 0:  # pick up last rejection run
+        autocorrelations[1 : consecutive_rejections + 1] += 1
+
+    # Compute integrated autocorrelation
+    integrated_autocorrelation = 0.5 + torch.sum(
+        autocorrelations / (N - torch.arange(N - 1))
+    )
```

Comment:
I think there are some missing counts when incrementing the autocorrelation slice, because autocorrelations[x] when x<consecutive_rejections must be incremented with a multiplicity... for example, inside consecutive_rejections=3 there are two combinations that give consecutive_rejections=2, "(0 0) 0" and "0 (0 0)".
A check for the implementation is to verify that in case of all rejections the autocorrelation function is identically 1 (see eq 16)

I don't know if I can push a new commit in this request, but a fix could be:
```ruby
    N = len(history)
    autocorrelations = torch.zeros(N, dtype=torch.float)
    consecutive_rejections = 0

    for step in history[1 : ]:
        if step == True:
            if consecutive_rejections > 0:
                autocorrelations[1 : consecutive_rejections + 1] += torch.arange(consecutive_rejections, 0, -1, dtype=torch.float)
                consecutive_rejections = 0
        else:
            consecutive_rejections += 1
    if consecutive_rejections > 0:
        autocorrelations[1 : consecutive_rejections + 1] += torch.arange(consecutive_rejections, 0, -1, dtype=torch.float)

    integrated_autocorrelation = 0.5 + torch.sum(autocorrelations / torch.arange(N, 0, -1, dtype=torch.float))
```

Reply:
Good spot! Thanks!